File: it4all-client/src/app/_interfaces/lesson.ts
type TutorialContentType = 'Text' | 'TrueFalseQuestion' | 'Questions' | 'Definition';

export interface LessonContentBase {
  _type: TutorialContentType;
  id: number;
  lessonId: number;
  toolId: string;
  priorSolved?: boolean;
}

export interface LessonTextContent extends LessonContentBase {
  _type: 'Text';
  content: string;
}

export interface LessonAnswer {
  answer: string;
  isCorrect: boolean;
}

export interface LessonQuestion {
  id: number;
  question: string;
  answers: LessonAnswer[];
}

export interface LessonQuestionsContent extends LessonContentBase {
  _type: 'Questions';
  questions: LessonQuestion[];
}

export interface LessonTrueFalseQuestion extends LessonContentBase {
  _type: 'TrueFalseQuestion';
  question: string;
  isTrue: boolean;
}

export function isLessonTextContent(content: LessonContentBase): content is LessonTextContent {
  return content._type === 'Text';
}

type LessonContent = LessonTextContent | LessonQuestionsContent | LessonTrueFalseQuestion;

/**
 * @deprecated
 */
export interface Lesson {
  id: number;
  toolId: string;
  title: string;
  description: string;
  content: LessonContent[];
  dependsOn?: number[];
}

File: app/model/lesson/Lesson.scala
package model.lesson

sealed trait LessonContent {
  val contentId: Int
  val lessonId: Int
  val toolId: String
}

final case class LessonTextContent(
  contentId: Int,
  lessonId: Int,
  toolId: String,
  content: String
) extends LessonContent

final case class LessonMultipleChoiceQuestionAnswer(
  answer: String,
  isCorrect: Boolean
)

final case class LessonMultipleChoiceQuestion(
  id: Int,
  question: String,
  answers: Seq[LessonMultipleChoiceQuestionAnswer]
)

final case class LessonMultipleChoiceQuestionsContent(
  contentId: Int,
  lessonId: Int,
  toolId: String,
  questions: Seq[LessonMultipleChoiceQuestion]
) extends LessonContent

final case class Lesson(
  lessonId: Int,
  toolId: String,
  title: String,
  description: String
)
